pkg/controller/backup_storage_location_controller.go

```diff
@@ -0,0 +1,210 @@
+		backupStorageLocationLister:  backupStorageLocationLister,
+
+		// use variables to refer to these functions so they can be
+		// replaced with fakes for testing.
+		newPluginManager: newPluginManager,
+		newBackupStore:   persistence.NewObjectBackupStore,
+	}
+
+	c.resyncFunc = c.run
+	c.resyncPeriod = 30 * time.Second
+
+	return c
+}
+
+func (c *backupStorageLocationController) run() {
+	c.logger.Info("Checking for existing locations ready to be validated; there needs to be at least 1 location that is available")
+
+	locations, err := c.backupStorageLocationLister.BackupStorageLocations(c.namespace).List(labels.Everything())
+	if err != nil {
+		c.logger.WithError(err).Error("Error listing locations, at least one available location is required")
+		return
+	}
+
+	if len(locations) == 0 {
+		c.logger.Error("No locations found, at least one available location is required")
+		return
+	}
+
+	pluginManager := c.newPluginManager(c.logger)
+	defer pluginManager.CleanupClients()
+
+	var unavailable []string
+	log := c.logger
+	for _, location := range locations {
+		locationName := location.Name
+		log = c.logger.WithField("backupLocation", locationName)
+
+		storeValidationPeriod := c.defaultStoreValidationPeriod
+		if location.Spec.StoreValidationPeriod != nil {
+			storeValidationPeriod = location.Spec.StoreValidationPeriod.Duration
+			if storeValidationPeriod == 0 {
+				log.Debug("Validation period for this location is set to 0, skipping validation")
+				continue
+			}
+
+			if storeValidationPeriod < 0 {
+				log.Debug("Validation period must be non-negative")
+				storeValidationPeriod = c.defaultStoreValidationPeriod
+			}
+		}
+
+		lastValidation := location.Status.LastValidationTime
+		if lastValidation != nil {
+			log.Debug("Checking if location needs to be validated at this time")
+			nextValidation := lastValidation.Add(storeValidationPeriod)
+			if time.Now().UTC().Before(nextValidation) {
+				continue
+			}
+		}
+
+		log.Debug("Location is ready to be validated")
+
+		backupStore, err := c.newBackupStore(location, pluginManager, log)
+		if err != nil {
+			unavailable = append(unavailable, errors.Wrapf(err, "error getting backup store for location %q", locationName).Error())
+			continue
+		}
+
+		if err := backupStore.IsValid(); err != nil {
+			// update status
+			err2 := c.patchBackupStorageLocation(location, func(r *velerov1api.BackupStorageLocation) {
+				r.Status.Phase = velerov1api.BackupStorageLocationPhaseUnavailable
+				r.Status.LastValidationTime = &metav1.Time{Time: time.Now().UTC()}
+			})
+			if err2 != nil {
+				unavailable = append(unavailable, errors.Wrapf(err2, "error updating location %q to %s status", locationName, velerov1api.BackupStorageLocationPhaseUnavailable).Error())
+			} else {
+				unavailable = append(unavailable, errors.Wrapf(err, "location %q is unavailable", locationName).Error())
+			}
+			log.Debug("Location has been updated to Unavailable")
```

Comment:
Suggest making this `log.Errorf` and also logging `err`

Reply:
```suggestion
			log.Errorf("BSL marked as unavailable, reason: %v", err)
```

Reply:
I removed it actually, now that this block is cleaner I think I was double logging in the debug case.